crates/flycheck/src/project_json.rs

```diff
@@ -0,0 +1,155 @@
+    }
+}
+
+/// A handle to a spawned [Discover].
+#[derive(Debug)]
+pub struct DiscoverHandle {
+    _handle: CommandHandle<DiscoverProjectMessage>,
+}
+
+/// An enum containing either progress messages or the materialized rust-project.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(tag = "kind")]
+#[serde(rename_all = "snake_case")]
+enum DiscoverProjectData {
+    Finished { buildfile: Utf8PathBuf, project: ProjectJsonData },
+    Error { error: String, source: Option<String> },
+    Progress { message: String },
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum DiscoverProjectMessage {
+    Finished { project: ProjectJsonData, buildfile: AbsPathBuf },
+    Error { error: String, source: Option<String> },
+    Progress { message: String },
+}
+
+impl DiscoverProjectMessage {
+    fn new(data: DiscoverProjectData) -> Self {
+        match data {
+            DiscoverProjectData::Finished { project, buildfile, .. } => {
+                let buildfile = buildfile.try_into().expect("Unable to make path absolute");
+                DiscoverProjectMessage::Finished { project, buildfile }
+            }
+            DiscoverProjectData::Error { error, source } => {
+                DiscoverProjectMessage::Error { error, source }
+            }
+            DiscoverProjectData::Progress { message } => {
+                DiscoverProjectMessage::Progress { message }
+            }
+        }
+    }
+}
+
+impl ParseFromLine for DiscoverProjectMessage {
+    fn from_line(line: &str, _error: &mut String) -> Option<Self> {
+        // can the line even be deserialized as JSON?
+        let Ok(data) = serde_json::from_str::<Value>(line) else {
+            let err = DiscoverProjectData::Error { error: line.to_owned(), source: None };
+            return Some(DiscoverProjectMessage::new(err));
+        };
+
+        let Ok(data) = serde_json::from_value::<DiscoverProjectData>(data) else {
+            return None;
+        };
+
+        let msg = DiscoverProjectMessage::new(data);
+        Some(msg)
+    }
+
+    fn from_eof() -> Option<Self> {
+        None
+    }
+}
+
+#[test]
+fn test_deserialization() {
+    let message = r#"
+    {"kind": "progress", "message":"querying build system","input":{"files":["src/main.rs"]}}
+    "#;
+    let message: DiscoverProjectData =
+        serde_json::from_str(message).expect("Unable to deserialize message");
+    assert!(matches!(message, DiscoverProjectData::Progress { .. }));
+
+    let message = r#"
+    {"kind": "error", "error":"failed to deserialize command output","source":"command"}
+    "#;
+
+    let message: DiscoverProjectData =
+        serde_json::from_str(message).expect("Unable to deserialize message");
+    assert!(matches!(message, DiscoverProjectData::Error { .. }));
+
+    let message = r#"
+    {"kind": "finished", "project": {"sysroot": "foo", "crates": [], "runnables": []}, "buildfile":"/Users/dbarsky/Developer/rust-analyzer"}
```

Comment:
>  Should buildfile be called root_dir instead?

No, I botched the copy-and-paste job for this test. It should be something like `Users/dbarsky/Developer/rust-analyzer/BUCK`.

> For reviewing, it flows through: [redacted steps]

This is correct, yes.

Reply:
@cormacrelf Sorry, I forgot to respond to this:

> Forcing RP to emit a buildfile seems a bit unnecessarily limiting, assumes you want one BUCK file at a time as your RA workspace and can somewhat haphazardly/heuristically add extra targets. What if I want to make RP support rust-project develop-json -t //... $JSON to make sure I don't end up with an important crate stuck as is_in_workspace: false? What should be the buildfile then?

I'm risking going off-topic about rust-project/Buck-specific concepts, but the change to rust-project would be to return a Vec of `DiscoverProjectData::Finished` instead of the current output. We'd know the set of related crates through by setting `with_rust_workspace` in a PACKAGE (not build!) file, which would behave like a Cargo workspace equivalent (note that we recently broke `with_rust_workspace`, so don't try to use it yet!).

Alternatively, you could have all your crates in a single `BUCK` file, but that loses the nice 1:1 correspondence between a Cargo.toml file and a `BUCK` file.

Reply:
What would RA even do with two separate json projects with about 600 dependencies in common?

Reply:
> What would RA even do with two separate json projects with about 600 dependencies in common?

It'll index them both and not perform and deduplication in terms of work or memory. This isn't great, but I think the correct fix is to make the crate graph an interned Salsa DB, which I'll be doing next. The state you're describing will likely be possible for a little while.

Reply:
My instinct here is to leverage Buck a bit more, and let it do the deduplication. It's really good at that. rust-project already supports passing in multiple target expressions and that's the main way I use it. So I would want RA to be able to drive it that way, and if it can't, I'll probably end up submitting/carrying a patch that allows overriding the targets for the develop-json subcommand, to ignore what RA says it's interested in, but still use the automatic discovery feature to reload.

A good interface for users to describe by their behaviour which targets to root the graph on is hard to pin down, and I can see why it'll take more than this PR to figure it out. But it doesn't matter too much to me because I have the luxury of a small enough repo where //... is almost usable.

Reply:
> So I would want RA to be able to drive it that way, and if it can't, I'll probably end up submitting/carrying a patch that allows overriding the targets for the develop-json subcommand, to ignore what RA says it's interested in, but still use the automatic discovery feature to reload.

At least on the rust-analyzer side of things, I _think_ that it should be pretty easy to support `//...` if `{arg}` is no longer required—rust-analyzer will still pass around the path to buildfile or Rust source file, but it'll never actually be _used_ if `{arg}` isn't, well, used. The tricky implementation bit would be on the `rust-project` side of things, but if I'm not mistaken, a regular `rust-project develop --stdout //...` might also _just work_ (but without any of the progress reporting).

I can imagine an alternative design where `{arg}` is still required, but it allows specifying literals like `{arg: '//...'}`, which would be used unconditionally, which would the behavior a bit more specified.

You know, I'll think I'll remove the hard requirement of specifying `{arg}` for now.